### PR TITLE
fix(dagster): bound failure retries and make the failure signal legible

### PR DESCRIPTION
**Pull request by blarghmatey**
## What are the relevant tickets?

Advances `wp-dagster-failure-semantics-sentry-signal-quality-f52032`. Closes six tasks:

- `tk-bound-the-failure-retry-term-in-upstream-or-code-3097b0` (p0) — M1
- `tk-give-permanent-failure-teeth-at-the-run-level-5454f8` (p0) — M2
- `tk-make-fileobjectiomanager-verify-an-object-exists-42bee5` (p1)
- `tk-scope-iceberg-maintenance-to-dagster-env-qa-is-t-5ba48f` (p1)
- `tk-have-the-airbyte-assets-check-for-an-in-flight-s-3ce4c0` (p2)
- `tk-make-the-sentry-pipeline-report-defects-not-laun-985529` (p1, 3 of 4 parts — see below)

Fixes DAGSTER-E. Partially addresses DAGSTER-1C..1V, DAGSTER-6, DAGSTER-13, DAGSTER-R, and the ten Airbyte "already running" issues.

## Description (What does it do?)

A permanent failure had no way to stop being re-run, and the reporting pipeline could not tell you that was happening. ~368,000 failed runs from one asset in fourteen days, reported as 27.9k Slack messages nobody could read.

### M1 — unbounded failure retry

`upstream_or_code_changes()` ORed in a bare `AutomationCondition.execution_failed()`. That term is level-triggered — true for as long as the latest execution is a failure — so a partition that can never succeed was re-requested on every sensor tick forever, and `run_retries.max_retries: 3` multiplied each request by four. `.newly_true()` spends one failure on one re-request: the level never drops back to false while the retry is in flight or after it also fails, and a success clears it so the next genuine failure earns its own. One function, every code location — this caps Group A (395,895 events) and Group B (8,180) together.

I checked `.since(newly_requested())` first, which the task suggested: it does not bound anything. `SinceCondition` computes `previous_true ∪ trigger_true − reset_true`, so a level-triggered trigger re-adds itself every tick after the reset clears. The edge has to come first.

### M2 — "permanent" never reached the rerun layer

Dagster has three independent retry mechanisms and `Failure(allow_retries=False)` speaks to exactly one. An op's `RetryPolicy` honours it; the run-level auto-reexecution daemon decides purely from run status and the `dagster/retry_on_asset_or_op_failure` tag and never inspects the exception; an `AutomationCondition` sees only that the latest execution failed. `http_errors.py` documented this gap and nothing closed it, so the Learn API 405 was correctly classified as unretryable and then retried 5,363 times over six days.

New `ol_orchestrate.lib.failures` carries a `PermanentFailure` base and a `stop_run_retries` failure hook that stamps the tag the daemon reads. `PermanentHTTPFailure` and `UpstreamObjectUnavailable` inherit from it. `with_sentry_hooks` is superseded by `with_failure_hooks`, which attaches both; all eight code locations move over and were import-checked individually.

A hook rather than a call at each raise site: the tag has to be set for every permanent failure or the guarantee is only as good as whoever remembered, and the raise sites do not all have an op context to hand.

**The third gap stays open by construction**, and the module docstring says so. Declarative automation exposes no primitive for "the failure was permanent" — `execution_failed()` is the only failure signal and it is untyped. The M1 bound is what covers it: a permanent failure now costs two runs rather than an unbounded stream.

Bare `raise Exception` sites migrated, which were never classified at all:

- edxorg Learn API webhook → `http_failure()`, so a 405 is permanent and a 502 still retries; the course id moves out of the message into metadata.
- openedx Studio export reporting `Failed` → `permanent_failure()` (DAGSTER-6, 2,587 events).
- canvas export in `failed`/`error` state → `permanent_failure()`.
- canvas export timeout → a plain `Failure`, **deliberately not permanent**. A slow export can genuinely succeed later; DAGSTER-7's problem was that the retry was unbounded, not that it retried. Claiming permanence here would suppress a real retry.
- lakehouse snapshot repair → `Failure` with the table list in metadata rather than the message, where it was making each occurrence its own issue title.

### Group A's actual defect

`FileObjectIOManager.load_input` trusted the materialization record unconditionally: it indexed `[0]` into the event records, read `metadata["path"]` with no guard, and handed back a path to an object that was not in S3. The `NoSuchKey` then surfaced deep inside whatever library opened it, naming nothing. It now checks the record exists, carries `path` metadata, and points at something present, and raises a non-retryable `UpstreamObjectUnavailable` naming the missing key.

### DAGSTER-R — Iceberg maintenance scoping

`load_maintenance_configs_from_manifest` read `node["schema"]` out of a manifest compiled against production and baked into an image that ships to every environment, so QA was issuing `OPTIMIZE`/`ANALYZE` against `ol_warehouse_production_intermediate`. The only thing stopping it was `data-lake-query-engine-role-qa` lacking `glue:GetTable`. Schemas are now re-pointed at `DAGSTER_ENV`'s warehouse, and a schema that is not `ol_warehouse_<env>_<layer>` raises rather than passing through. The raw-layer map and the hand-written singleton list both moved onto the same mapping so they cannot drift.

To be fair to the IAM policy: that denial is deliberate, not accidental. Both identities that reach Glue — the Starburst cross-account role and the Dagster pod IRSA role — scope resources to `arn:aws:glue:*:*:database/*{env_suffix}*` in Pulumi, and have since 2022. IAM is a working backstop here. It is still the wrong layer to be relying on, because it tells you "access denied" rather than "this should never have been attempted", and because the pattern is a substring match (see below).

Also fixed the arithmetically impossible threshold message — `failures` counted operations while `tables_processed` counted tables, producing "failed for 1256/628 tables". Both sides now count tables, and the denominator is tables *attempted* so a total outage does not divide by zero and silently never trip.

### M3 — the reporting pipeline

The Sentry fingerprint led with `job_name`, which describes how a run was launched rather than what broke: DAGSTER-2C and DAGSTER-29 were one OVS webhook failure split only because one run came from `__ASSET_JOB` and the other from `ovs_videos_webhook_job`. It is now `[environment, code_location, step_key, exception_type]`, built by one shared `failure_fingerprint()` so the in-process hook and the run failure sensor cannot drift apart. The environment leads because QA and production share a Sentry project. The sensor was also tagging every run's code location as `"data_platform"` regardless of origin; it now reads the run's actual origin.

Slack gets one message per distinct failure per 30 minutes, carrying a count of what it swallowed. Sentry still gets every event — it aggregates, and the count there is the signal. Rate-limit state lives in the run storage KV table because `RunFailureSensorContext` is invoked once per failed run and exposes no cursor.

`DagsterExecutionInterruptedError` and `KeyboardInterrupt` are dropped in a `before_send` (DAGSTER-1R/1S/1T).

### Two defects that turned out to be different from their descriptions

**Airbyte** — the cause was not a missing in-flight check. `dagster_airbyte.sync_and_poll` already checks, and has a `poll_previous_running_sync` flag to attach to a running sync instead of raising. The real bug is that `AirbyteOSSWorkspace.get_client()` re-implements the base class's constructor call with an explicit argument list, and four settings were missing from it: `poll_interval`, `poll_timeout`, `cancel_on_termination` and `poll_previous_running_sync` (plus `workspace_id`). Configuring any of them on the workspace set a field the client never read, so every sync ran on library defaults — and the flag defaults to False, hence ten connections raising "already running". Fixed by passing them through and turning the flag on.

**edxorg CSV** — the dialect is already pinned; that is not the fix. `delimiter`, `quotechar`, `escapechar`, `strict_mode` and `all_varchar` all landed in #2492 on 2026-07-30. The DAGSTER-1C..1V events are 2026-08-13 to 08-16 on release 2b127d70, and the Sentry error text shows the pinned dialect in use (`Delimiter Candidates: '\t'`, `Quote/Escape Candidates: ['"','"']`). Pinning the dialect does not remove the sniffer — DuckDB still runs it to find the header and column count.

What actually blocks diagnosis is that `from_csv_auto` is handed an open file object, so the file it names is DuckDB's internal handle, `DUCKDB_INTERNAL_OBJECTSTORE://e3d60147029d6cb5`, which maps to nothing anyone can open. `read_edxorg_tsv` does the same work wrapped to put the S3 URL in the exception, and to skip zero-byte files, which fail the sniffer with the identical message despite not being errors. **This does not close the underlying bug** — it makes the next occurrence reproducible.

### DAGSTER-E — two bugs in one traceback

The reported one: `headers.get("Retry-After", 60)` defaults to an *int*, so `retry_after.isdigit()` raised `AttributeError` and the 429 retry path crashed on a 429.

The one underneath it, visible in the URL that got rate limited: `parse_qs` was applied to the API's absolute `next` URL. `parse_qs` on a full URL takes everything up to the first `=` as a key, so `https://host/courses/?page=2` parsed to `{"https://host/courses/?page": ["2"]}` — and that key was sent as a query parameter *name*. The server echoed the mangled parameters back into the next `next`, so each page nested the base URL one level deeper; by page 27 the request contained its own base URL twenty-five times over and edX answered 429. That was not the rate limit doing its job, it was one client hammering an endpoint with a URL that grew every round. Three pagination loops had it.

## Deploy note — read before rolling this out

`NewlyTrueCondition` diffs the child's true subset against the previous tick's subset held in its cursor. On the first evaluation after deploy there is no cursor, so **every currently-failed partition reads as newly true and fires one request each, on one tick**. Against the ~368k standing failed partitions that is the exact stampede this change exists to prevent, once.

`initial_evaluation()` does not guard against this — I checked the implementation; it keys off whether a previous requested subset exists for the root key (and off `PartitionsDefinition` changes), not off the condition tree changing, so it is already false for any asset evaluated before. There is no in-condition defence; the constraint is a deployment one.

Land the edxorg backfill stop and clear the failed set first, or roll this out one code location at a time starting with small partition counts. This is documented in the function's docstring as well.

## How can this be tested?

All green: `packages/ol-orchestrate-lib` 156 passed / 80 skipped, `data_platform` 52, `lakehouse` 78, `openedx` 25, `data_loading` 18, `ol_dlt` 90. `pre-commit run` clean including mypy. All eight code locations import-checked individually after the `with_failure_hooks` migration.

New coverage worth looking at specifically:

- `test_the_failure_retry_is_one_run_not_a_treadmill` and `test_a_later_failure_gets_its_own_retry` — the bound holds *and* does not degrade to "one retry per asset, ever". The existing `test_update_is_retried_after_a_failed_run` (the mitodl/hq#12739 regression) still passes, so the property the docstring was protecting survives.
- `tests/lib/test_failures.py` — drives real runs and asserts on the run tag the daemon reads, including that plain `Failure`, `TransientHTTPFailure` and `RuntimeError` are **not** tagged. Over-tagging would silently disable retries for every ordinary failure.
- `tests/io_managers/test_filepath.py` — all three IO manager failure modes against a real local filesystem.
- `test_a_qa_load_never_returns_a_production_schema` — the DAGSTER-R bug directly.
- `lakehouse_tests/test_airbyte_resource.py` — pins every workspace setting individually, so the next edit to that argument list cannot quietly drop another.
- `test_sensor_fingerprint_matches_a_real_dagster_step_failure` still executes a real job and compares against what the actual hook records, so the two paths cannot drift silently.

## Additional context

Left for follow-up, deliberately, and **not** closed:

- **Identifiers in exception messages** (M3 part 4). Partially done — the edxorg course id, the Canvas course id, the Studio status map and the lakehouse table list all moved to metadata — but `load_id=...`, video ids and Airbyte connection ids remain in message text elsewhere.
- **Why the sensor reported only 27.9k of 368k failures.** Retry-dedup accounts for at most 75%; the sensor appears to be falling behind its own tick. Untouched.
- **`RetryPolicy` on the DB/HTTP-touching assets.** The `before_send` here is deliberately narrow: it drops terminated workers only. Connection-class transients should be *retried*, not filtered — a step failure hook fires only once a `RetryPolicy` is exhausted, so a connection error that still reaches Sentry is one that survived its retries and is worth reading.
- **The edxorg CSV sniffing bug itself**, per above.
- **DAGSTER-9, DAGSTER-M, DAGSTER-F** — the three remaining low-volume defects.

**Checked, so no longer an open question:** the QA role never had broader Glue access. It has been environment-scoped since it was created in 2022 (mitodl/ol-infrastructure#869), originally as a suffix match `database/*qa`, widened to the substring form `database/*qa*` on 2022-09-28 (`5c2886030`) so dbt's `ol_warehouse_qa_<layer>` names — which carry the env in the middle — would match. The only later change added a userDefinedFunction ARN.

**A separate finding, now fixed in mitodl/ol-infrastructure#5472.** That substring match means `*qa*` reaches any database whose name contains "qa" anywhere, and five in the production namespace do: `ol_warehouse_production_qa_{external,intermediate,mart,staging}` and `ol_warehouse_mitx_qa`. `ol_warehouse_production_qa_mart` is registered at `s3://ol-data-lake-staging-production/...`. All five are vestigial dbt developer schemas from 2023-07-28 with zero tables, so nothing is exposed today — but the granted actions include `glue:DeleteTable` and `glue:DeleteDatabase`, and this becomes real the moment a production schema containing "qa" holds data. The S3 grant is scoped tighter (`ol-data-lake-*-{env_suffix}`, a trailing match), so the exposure is metadata-only. mitodl/ol-infrastructure#5472 anchors the pattern to the real naming conventions and adds an explicit Deny on the production namespace for the non-production stacks; it is strictly narrowing, verified against the live catalog, and previews clean on all six affected stacks.

The pgbouncer churn underneath Group H is a symptom already scoped under `wp-dagster-pgbouncer-observability-and-pool-tuning-cce725` and is not re-scoped.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PK9Bt1uMHXLysq6KuenDUR

### CODE REVIEW

File: packages/ol-orchestrate-lib/tests/lib/test_sentry.py [hunk 1]
```diff
@@ -158,10 +159,43 @@ def exploding_asset() -> None:
...
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """One Sentry project holds both, so the environment has to group them apart.
+
+    Without it a QA-only defect and a production outage merge into one issue and
+    the list cannot tell you which you are looking at.
+    """
+    production = sentry_lib.failure_fingerprint("lakehouse", "dbt_build", "Failure")
+    monkeypatch.setattr(sentry_lib, "DAGSTER_ENV", "qa")
+    qa = sentry_lib.failure_fingerprint("lakehouse", "dbt_build", "Failure")
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply packages/ol-orchestrate-lib/tests/lib/test_sentry.py:187]: Fixed in 0e1ba89a: both environments are now set explicitly, and each fingerprint asserts its own env. It was a tautology under `ci`/`dev` and an outright failure under `qa`.

File: packages/ol-orchestrate-lib/src/ol_orchestrate/lib/sentry.py [hunk 2]
```diff
@@ -36,6 +36,47 @@
...
+    Deliberately narrow. Connection-class transients -- pgbouncer churn, httpx
+    timeouts, DNS -- are handled by giving the assets that touch the database
+    and external HTTP a ``RetryPolicy`` instead of by filtering here: a step
+    failure hook fires only once a RetryPolicy is exhausted, so a connection
+    error that still reaches Sentry is one that survived its retries and is
+    worth reading. Filtering those at the transport would hide the sustained
+    outage along with the blip.
+    """
+    if exception_type_of(event) in INTERRUPTION_ERRORS:
+        return None
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply packages/ol-orchestrate-lib/src/ol_orchestrate/lib/sentry.py:72]: Fixed in 0e1ba89a. Correct, and it is the same two-capture-paths-must-agree trap this PR fixes for fingerprints: `capture_message` events carry no `exception.values`, so an interrupted worker was dropped on the hook path and reported on the sensor path. Added `is_an_interrupted_worker`, which reads the serialized step error (where the type survives) and unwraps Dagster's wrapper first.

File: dg_projects/data_platform/data_platform/definitions.py [hunk 3]
```diff
@@ -348,6 +380,56 @@ def format_error(error_event):
...
+    unbatched stream rendered them identically.
+
+    State lives in the run storage KV table rather than the sensor cursor:
+    ``RunFailureSensorContext`` is invoked once per failed run and exposes no
+    cursor to carry counts between those invocations.
+    """
+    record = instance.run_storage.get_cursor_values({key}).get(key)
+    if not record:
+        instance.run_storage.set_cursor_values({key: f"{now}:0"})
+        return 0
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply dg_projects/data_platform/data_platform/definitions.py:452]: Fixed in 0e1ba89a: split into `repeats_to_announce` (decide) and `record_announcement` (commit, after the post returns). A Slack or Vault failure would have silenced alerting for 30 minutes instead of shouting about it, which inverts the point of leaving the post unguarded. Suppression counts are still written inline, since that path posts nothing.

File: dg_projects/data_platform/data_platform/definitions.py [hunk 4]
```diff
@@ -384,15 +467,34 @@ def run_failure_notification_sensor(context: RunFailureSensorContext) -> None:
...
     sentry_event_id = capture_run_failure_to_sentry(context)
 
+    # Sentry still receives every failure -- it aggregates, and the count is the
+    # signal. Slack does not aggregate, so the same failure repeating is rate
+    # limited to one message per window with a count of what it swallowed.
+    repeats = repeats_to_announce(
+        context.instance,
+        slack_announcement_key(fingerprint),
+        context.dagster_run.create_timestamp.timestamp(),
+    )
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply dg_projects/data_platform/data_platform/definitions.py]: Fixed in 0e1ba89a: uses `datetime.now(tz=UTC)`. A backlog drained in one tick carries creation timestamps hours apart and out of order, so it could both emit several messages at once and suppress newer failures behind older ones.

File: dg_projects/lakehouse/lakehouse/assets/iceberg_maintenance.py [hunk 5]
```diff
@@ -287,41 +295,48 @@ def iceberg_dbt_layer_maintenance(
...
+    # A table that failed every operation is one table, not three: the ratio is
+    # only meaningful if both sides count the same thing. The denominator is
+    # tables attempted rather than `tables_processed`, which excludes the ones
+    # that raised outright -- under a total outage that is zero, and a threshold
+    # against zero never trips.
     if failures:
-        failure_threshold = max(1, int(tables_processed * 0.05))
-        if len(failures) >= failure_threshold:
+        failure_threshold = max(1, int(tables_attempted * 0.05))
+        if len(failed_tables) >= failure_threshold:
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply dg_projects/lakehouse/lakehouse/assets/iceberg_maintenance.py]: Fixed in 0e1ba89a: `floor(5%) + 1`, which also makes the `max(1, ...)` floor fall out naturally. Extracted to `maintenance_failure_threshold` in ol_orchestrate.lib with your two boundary cases (1/21 and 5/100) pinned as tests.

**Comment by blarghmatey** [comment]
Addressed all five Copilot findings in 0e1ba89a — threads replied to and resolved.

Two were substantive: the interruption filter never covered the sensor's `capture_message` path (so a SIGTERM'd worker was dropped on the hook path and reported on the other one), and the Slack rate limiter opened its window *before* posting, so a Slack or Vault outage would have silenced alerting for 30 minutes instead of failing loudly. The other three were a wall-clock-vs-run-creation-time bug in the same limiter, a test that asserted its property only by accident, and a threshold that didn't implement the "more than 5%" it documented.

All suites green.
